### PR TITLE
cmath: start using it

### PR DESCRIPTION
**Pull request by ricardotk**
These series swaps math.h usage to cmath usage.

It all started with an update on ArchLinux that broke ArduPilot build for me, the fix was to put `std::` in front of `isnan()`, `isinf()` and `isfinite()`. But I decided to send a better fix here.

Thanks!

### CODE REVIEW

**Comment by magicrub** [comment]
@dgrat has been working on a math library upgrade as well, you guys should talk. Please see his 5th iteration on his huge PR https://github.com/ArduPilot/ardupilot/pull/3664

**Comment by dgrat** [comment]
cmath is increasing the flash size

**Comment by ricardotk** [comment]
@magicrub yes, I've already looked at PR #3664 but mine is more like a hot fix for the build.

**Comment by ricardotk** [comment]
@dgrat I had to put `std::` in front of `isnan()`, `isinf()` and `isfinite()` after `gcc`, `g++` and `glibc` update in order to build it without errors. Any suggestions on how to fix it without increasing the flash size?

**Comment by lucasdemarchi** [comment]
I mentioned the PR from @dgrat as a way to fix it, but apparently he's not handling these cases... particularly because these are not about the PX4 toolchain. It's about different linux toolchains.

@dgrat any other way to fix this?

@ricardotk we should detect in the build system if that's supported by the toolchain. Take a look in check_cxx(https://waf.io/book/#_configuration_helpers).  I think that writing the HAVE_XXX to a config.h (which we should give another name due to name conflict, maybe toolchain.h) and then pass as -include to the compiler would be the right thing to do instead of polluting even more with -D flags

File: libraries/AP_Common/missing/cmath [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include_next <cmath>
+
+#if defined(_GLIBCXX_CMATH)
```
Comment (lucasdemarchi) [inline libraries/AP_Common/missing/cmath]: this is actually wrong... a) it's a "private" guard used by glibcxx... not to be used here; b) we should ifdef with what the build system detected; c) I think this fails if they are not undef'ed by the header, i.e. it fails on !g++ and g++ < 5.3

**Comment by ricardotk** [comment]
@lucasdemarchi Yes, that would be a better solution (: thanks

**Comment by ricardotk** [comment]
@dgrat @lucasdemarchi 

Actually using cmath didn't increased the flash size.

| text | data | bss | dec | hex | filename
|----- | ------ | ---- | ----- | ----- | ---------- |
|972588    |2820   |57116 |1032524 |  fc14c | master

| text | data | bss | dec | hex | filename
|----- | ------ | ---- | ----- | ----- | ---------- |
| 972628   | 2820   | 57132 | 1032580 |  fc184 | cmath

**Comment by ricardotk** [comment]
v2 pushed #3833

**Pull request closed without merging**